server/src/com/thoughtworks/go/config/update/PatchEnvironmentCommand.java

```diff
@@ -0,0 +1,141 @@
+            }
+        }
+
+        if (!unknownAgents.isEmpty()) {
+            result.badRequest(LocalizedMessage.string("AGENTS_WITH_UUIDS_NOT_FOUND", unknownAgents));
+            return false;
+        }
+        return true;
+    }
+
+    private boolean validatePipelines(List<String> pipelines, HttpLocalizedOperationResult result, CruiseConfig config) {
+        ArrayList<String> unknownPipelines = new ArrayList<>();
+
+        for (String pipeline : pipelines) {
+            try{
+                config.pipelineConfigByName(new CaseInsensitiveString(pipeline));
+            } catch(PipelineNotFoundException e){
+                unknownPipelines.add(pipeline);
+            }
+        }
+
+        if (!unknownPipelines.isEmpty()) {
+            result.badRequest(LocalizedMessage.string("PIPELINES_WITH_NAMES_NOT_FOUND", unknownPipelines));
+            return false;
+        }
+        return true;
+
+    }
+
+
+    @Override
+    public void update(CruiseConfig preprocessedConfig) throws Exception {
+        EnvironmentsConfig environments = preprocessedConfig.getEnvironments();
+        int index = environments.indexOf(environmentConfig);
+        EnvironmentConfig preprocessedEnvironmentConfig = environments.get(index);
+
+        if(isValidConfig(preprocessedConfig)){
+            for (String uuid : agentsToAdd) {
+                preprocessedEnvironmentConfig.addAgent(uuid);
+            }
+
+            for (String uuid : agentsToRemove) {
+                preprocessedEnvironmentConfig.removeAgent(uuid);
+            }
+
+            for (String pipelineName : pipelinesToAdd) {
+                preprocessedEnvironmentConfig.addPipeline(new CaseInsensitiveString(pipelineName));
+            }
+
+            for (String pipelineName : pipelinesToRemove) {
+                preprocessedEnvironmentConfig.removePipeline(new CaseInsensitiveString(pipelineName));
+            }
+        }
+    }
+
+    @Override
+    public void clearErrors() {
+        BasicCruiseConfig.clearErrors(environmentConfig);
+    }
+
+    @Override
+    public boolean canContinue(CruiseConfig cruiseConfig) {
+        if (!goConfigService.isAdministrator(username.getUsername())) {
+            Localizable noPermission = LocalizedMessage.string("NO_PERMISSION_TO_UPDATE_ENVIRONMENT", environmentConfig.name().toString(), username.getDisplayName());
+            result.unauthorized(noPermission, HealthStateType.unauthorised());
+            return false;
+        }
+        return true;
+    }
+
+    public boolean isValidConfig(CruiseConfig preprocessedConfig) {
+        boolean isValid = validateAgents(agentsToAdd, result, preprocessedConfig.agents());
+        isValid = isValid && validateAgents(agentsToRemove, result, preprocessedConfig.agents());
+        isValid = isValid && validatePipelines(pipelinesToAdd, result, preprocessedConfig);
+        isValid = isValid && validatePipelines(pipelinesToRemove, result, preprocessedConfig);
```

Comment:
Should invert the `isValid && ...` to make it `... && isValid` that'll make sure to validate everything. Or you'll just validate until the first failure.